File: helper/kernel.c
/* supermin-helper reimplementation in C.
 * Copyright (C) 2009-2013 Red Hat Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <fnmatch.h>
#include <unistd.h>
#include <errno.h>
#include <sys/utsname.h>

#include "error.h"
#include "xvasprintf.h"
#include "full-write.h"

#include "helper.h"

#ifndef O_CLOEXEC
#define O_CLOEXEC 0
#endif

/* Directory containing candidate kernels.  We could make this
 * configurable at some point.
 */
#define KERNELDIR "/boot"
#define MODULESDIR "/lib/modules"

static char* get_kernel_version (char* filename);
static const char *create_kernel_from_env (const char *hostcpu, const char *kernel, const char *kernel_env, const char *modpath_env);
static void copy_or_symlink_kernel (const char *from, const char *to);

static char *
get_modpath (const char *kernel_name)
{
  /* Ignore "vmlinuz-" at the beginning of the kernel name. */
  const char *version = &kernel_name[8];

  /* /lib/modules/<version> */
  char *modpath = xasprintf (MODULESDIR "/%s", version);
  if (!modpath) {
    perror ("xasprintf");
    exit (EXIT_FAILURE);
  }

  if (! isdir (modpath)) {
    char* path;
    char* version;
    path = xasprintf (KERNELDIR "/%s", kernel_name);
    if (!path) {
      perror ("xasprintf");
      exit (EXIT_FAILURE);
    }
    version = get_kernel_version (path);
    free (path);
    if (version != NULL) {
      free (modpath);
      modpath = xasprintf (MODULESDIR "/%s", version);
      free (version);
      if (!path) {
        perror ("xasprintf");
        exit (EXIT_FAILURE);
      }
    }
  }

  return modpath;
}

/* kernel_name is "vmlinuz-*".  Check if there is a corresponding
 * module path in /lib/modules.
 */
static int
has_modpath (const char *kernel_name)
{
  char *modpath = get_modpath (kernel_name);

  if (verbose)
    fprintf (stderr, "checking modpath %s is a directory\n", modpath);

  int r = isdir (modpath);

  if (r) {
    free (modpath);
    return 1;
  }
  else {
    if (verbose)
      fprintf (stderr, "ignoring %s (no modpath %s)\n", kernel_name, modpath);
    free (modpath);
    return 0;
  }
}

/* Create the kernel.  This chooses an appropriate kernel and makes a
 * symlink to it (or copies it if --copy-kernel was passed).
 *
 * Look for the most recent kernel named vmlinuz-*.<arch>* which has a
 * corresponding directory in /lib/modules/. If the architecture is
 * x86, look for any x86 kernel.
 *
 * RHEL 5 didn't append the arch to the kernel name, so look for
 * kernels without arch second.
 *
 * If no suitable kernel can be found, exit with an error.
 *
 * This function returns the module path (ie. /lib/modules/<version>).
 */
const char *
create_kernel (const char *hostcpu, const char *kernel)
{
  /* Override kernel selection using environment variables? */
  char *kernel_env = getenv ("SUPERMIN_KERNEL");
  if (kernel_env) {
    char *modpath_env = getenv ("SUPERMIN_MODULES");
    return create_kernel_from_env (hostcpu, kernel, kernel_env, modpath_env);
  }

  /* In original: ls -1dvr /boot/vmlinuz-*.$arch* 2>/dev/null | grep -v xen */
  const char *patt;
  if (hostcpu[0] == 'i' && hostcpu[2] == '8' && hostcpu[3] == '6' &&
      hostcpu[4] == '\0')
    patt = "vmlinuz-*.i?86*";
  else
    patt = xasprintf ("vmlinuz-*.%s*", hostcpu);

  char **all_files = read_dir (KERNELDIR);
  char **candidates;
  candidates = filter_fnmatch (all_files, patt, FNM_NOESCAPE);
  candidates = filter_notmatching_substring (candidates, "xen");
  candidates = filter (candidates, has_modpath);

  if (candidates[0] == NULL) {
    /* In original: ls -1dvr /boot/vmlinuz-* 2>/dev/null | grep -v xen */
    patt = "vmlinuz-*";
    candidates = filter_fnmatch (all_files, patt, FNM_NOESCAPE);
    candidates = filter_notmatching_substring (candidates, "xen");
    candidates = filter (candidates, has_modpath);

    if (candidates[0] == NULL)
      goto no_kernels;
  }

  sort (candidates, reverse_filevercmp);

  if (verbose)
    fprintf (stderr, "picked %s\n", candidates[0]);

  if (kernel) {
    /* Choose the first candidate. */
    char *tmp = xasprintf (KERNELDIR "/%s", candidates[0]);
    copy_or_symlink_kernel (tmp, kernel);
    free (tmp);
  }

  return get_modpath (candidates[0]);

  /* Print more diagnostics here than the old script did. */
 no_kernels:
  fprintf (stderr,
           "supermin-helper: failed to find a suitable kernel.\n"
           "I looked for kernels in " KERNELDIR " and modules in " MODULESDIR
           ".\n"
           "If this is a Xen guest, and you only have Xen domU kernels\n"
           "installed, try installing a fullvirt kernel (only for\n"
           "supermin use, you shouldn't boot the Xen guest with it).\n");
  exit (EXIT_FAILURE);
}

/* Select the kernel from environment variables set by the user.
 * modpath_env may be NULL, in which case we attempt to work it out
 * from kernel_env.
 */
static const char *
create_kernel_from_env (const char *hostcpu, const char *kernel,
                        const char *kernel_env, const char *modpath_env)
{
  if (verbose) {
    fprintf (stderr,
             "supermin-helper: using environment variable(s) SUPERMIN_* to\n"
             "select kernel %s", kernel_env);
    if (modpath_env)
      fprintf (stderr, " and module path %s", modpath_env);
    fprintf (stderr, "\n");
  }

  if (!isfile (kernel_env)) {
    fprintf (stderr,
             "supermin-helper: %s: not a regular file\n"
             "(what is $SUPERMIN_KERNEL set to?)\n", kernel_env);
    exit (EXIT_FAILURE);
  }

  if (!modpath_env) {
    /* Try to guess modpath from kernel path. */
    const char *p = strrchr (kernel_env, '/');
    if (p) p++; else p = kernel_env;

    /* NB: We need the extra test to ensure calling get_modpath is safe. */
    if (strncmp (p, "vmlinuz-", 8) != 0) {
      fprintf (stderr,
               "supermin-helper: cannot guess module path.\n"
               "Set $SUPERMIN_MODULES to the modules directory corresponding to\n"
               "kernel %s, or unset $SUPERMIN_KERNEL to autoselect a kernel.\n",
               kernel_env);
      exit (EXIT_FAILURE);
    }

    modpath_env = get_modpath (p);
  }

  if (!isdir (modpath_env)) {
    fprintf (stderr,
             "supermin-helper: %s: not a directory\n"
             "(what is $SUPERMIN_MODULES set to?)\n", modpath_env);
    exit (EXIT_FAILURE);
  }

  /* Create the symlink. */
  if (kernel)
    copy_or_symlink_kernel (kernel_env, kernel);

  return modpath_env;
}

static void
copy_or_symlink_kernel (const char *from, const char *to)
{
  int fd1, fd2;
  char buf[BUFSIZ];
  ssize_t r;

  if (verbose >= 2)
    fprintf (stderr, "%s kernel %s -> %s\n",
             !copy_kernel ? "symlink" : "copy", from, to);

  if (!copy_kernel) {
    if (symlink (from, to) == -1)
      error (EXIT_FAILURE, errno, "creating kernel symlink %s %s", from, to);
  }
  else {
    fd1 = open (from, O_RDONLY | O_CLOEXEC);
    if (fd1 == -1)
      error (EXIT_FAILURE, errno, "open: %s", from);

    fd2 = open (to, O_WRONLY|O_CREAT|O_NOCTTY|O_CLOEXEC, 0644);
    if (fd2 == -1)
      error (EXIT_FAILURE, errno, "open: %s", to);

    while ((r = read (fd1, buf, sizeof buf)) > 0) {
      if (full_write (fd2, buf, r) != r)
        error (EXIT_FAILURE, errno, "write: %s", to);
    }

    if (r == -1)
      error (EXIT_FAILURE, errno, "read: %s", from);

    if (close (fd1) == -1)
      error (EXIT_FAILURE, errno, "close: %s", from);

    if (close (fd2) == -1)
      error (EXIT_FAILURE, errno, "close: %s", to);
  }
}

/* Read an unsigned little endian short at a specified offset in a file.
 * Returns a non-negative int on success or -1 on failure.
 */
static int
read_leshort (FILE* fp, int offset)
{
  char buf[2];
  if (fseek (fp, offset, SEEK_SET) != 0 ||
      fread (buf, sizeof(char), 2, fp) != 2)
  {
    return -1;
  }
  return ((buf[1] & 0xFF) << 8) | (buf[0] & 0xFF);
}

/* Extract the kernel version from a Linux kernel file.
 * Returns a malloc'd string containing the version or NULL if the
 * file can't be read, is not a Linux kernel, or the version can't
 * be found.
 *
 * See ftp://ftp.astron.com/pub/file/file-<ver>.tar.gz
 * (file-<ver>/magic/Magdir/linux) for the rules used to find the
 * version number:
 *   514             string  HdrS     Linux kernel
 *   >518            leshort >0x1ff
 *   >>(526.s+0x200) string  >\0      version %s,
 *
 * Bugs: probably limited to x86 kernels.
 */
static char*
get_kernel_version (char* filename)
{
  FILE* fp;
  int size = 132;
  char buf[size];
  int offset;

  fp = fopen (filename, "rb");

  if (fseek (fp, 514, SEEK_SET) != 0 ||
      fgets (buf, size, fp) == NULL ||
      strncmp (buf, "HdrS", 4) != 0 ||
      read_leshort (fp, 518) < 0x1FF)
  {
    /* not a Linux kernel */
    fclose (fp);
    return NULL;
  }

  offset = read_leshort (fp, 526);
  if (offset == -1)
  {
    /* can't read version offset */
    fclose (fp);
    return NULL;
  }

  if (fseek (fp, offset + 0x200, SEEK_SET) != 0 ||
      fgets (buf, size, fp) == NULL)
  {
    /* can't read version string */
    fclose (fp);
    return NULL;
  }

  fclose (fp);

  buf[strcspn (buf, " \t\n")] = '\0';
  return strdup (buf);
}

File: src/kernel.ml
(* supermin 5
 * Copyright (C) 2009-2014 Red Hat Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *)

open Unix
open Printf

open Utils
open Ext2fs
open Fnmatch

let rec build_kernel debug host_cpu dtb_wildcard copy_kernel kernel dtb =
  (* Locate the kernel. *)
  let kernel_name, kernel_version =
    find_kernel debug host_cpu copy_kernel kernel in

  (* If the user passed --dtb option, locate dtb. *)
  (match dtb_wildcard with
  | None -> ()
  | Some wildcard ->
    find_dtb debug copy_kernel kernel_name wildcard dtb
  );

  (* Get the kernel modules. *)
  let modpath = find_modpath debug kernel_version in

  if debug >= 1 then (
    printf "supermin: kernel: kernel_version %s\n" kernel_version;
    printf "supermin: kernel: modules %s\n%!" modpath;
  );

  (kernel_version, modpath)

and find_kernel debug host_cpu copy_kernel kernel =
  let kernel_file, kernel_name, kernel_version =
    try
      let kernel_env = getenv "SUPERMIN_KERNEL" in
      if debug >= 1 then
        printf "supermin: kernel: SUPERMIN_KERNEL environment variable %s\n%!"
          kernel_env;
      let kernel_version = get_kernel_version_from_file kernel_env in
      if debug >= 1 then
        printf "supermin: kernel: SUPERMIN_KERNEL version %s\n%!"
          kernel_version;
      let kernel_name = Filename.basename kernel_env in
      kernel_env, kernel_name, kernel_version
    with Not_found ->
      let is_x86 =
        String.length host_cpu = 4 &&
        host_cpu.[0] = 'i' && host_cpu.[2] = '8' && host_cpu.[3] = '6' in
      let is_arm =
        String.length host_cpu >= 3 &&
        host_cpu.[0] = 'a' && host_cpu.[1] = 'r' && host_cpu.[2] = 'm' in

      let all_files = Sys.readdir "/boot" in
      let all_files = Array.to_list all_files in

      (* In original: ls -1dvr /boot/vmlinuz-*.$arch* 2>/dev/null | grep -v xen *)
      let patt =
        if is_x86 then "vmlinuz-*.i?86*"
        else "vmlinuz-*." ^ host_cpu ^ "*" in
      let files = kernel_filter patt is_arm all_files in

      let files =
        if files <> [] then files
        else
          (* In original: ls -1dvr /boot/vmlinuz-* 2>/dev/null | grep -v xen *)
          kernel_filter "vmlinuz-*" is_arm all_files in

      if files = [] then no_kernels ();

      let files = List.sort (fun a b -> compare_version b a) files in
      let kernel_name = List.hd files in
      let kernel_version = get_kernel_version kernel_name in

      if debug >= 1 then
        printf "supermin: kernel: picked kernel %s\n%!" kernel_name;

      ("/boot" // kernel_name), kernel_name, kernel_version in

  copy_or_symlink_file copy_kernel kernel_file kernel;
  kernel_name, kernel_version

and kernel_filter patt is_arm all_files =
  let files =
    List.filter
      (fun filename -> fnmatch patt filename [FNM_NOESCAPE]) all_files in
  let files =
    List.filter (fun filename -> find filename "xen" = -1) files in
  let files =
    if not is_arm then files
    else (
      List.filter (fun filename ->
        find filename "lpae" = -1 && find filename "tegra" = -1
      ) files
    ) in
  List.filter (fun filename -> has_modpath filename) files

and no_kernels () =
  eprintf "\
supermin: failed to find a suitable kernel.

I looked for kernels in /boot and modules in /lib/modules.

If this is a Xen guest, and you only have Xen domU kernels
installed, try installing a fullvirt kernel (only for
supermin use, you shouldn't boot the Xen guest with it).\n";
  exit 1

and find_dtb debug copy_kernel kernel_name wildcard dtb =
  let dtb_file =
    try
      let dtb_file = getenv "SUPERMIN_DTB" in
      if debug >= 1 then
        printf "supermin: kernel: SUPERMIN_DTB environment variable = %s\n%!"
          dtb_file;
      dtb_file
    with Not_found ->
      (* Replace vmlinuz- with dtb- *)
      if not (string_prefix "vmlinuz-" kernel_name) then
        no_dtb_dir kernel_name;
      let dtb_dir =
        "/boot/dtb-" ^
          String.sub kernel_name 8 (String.length kernel_name - 8) in
      if not (dir_exists dtb_dir) then
        no_dtb_dir kernel_name;

      let all_files = Sys.readdir dtb_dir in
      let all_files = Array.to_list all_files in

      let files =
        List.filter (fun filename -> fnmatch wildcard filename [FNM_NOESCAPE])
          all_files in
      if files = [] then
        no_dtb dtb_dir wildcard;

      let dtb_name = List.hd files in
      let dtb_file = dtb_dir // dtb_name in
      if debug >= 1 then
        printf "supermin: kernel: picked dtb %s\n%!" dtb_file;
      dtb_file in

  copy_or_symlink_file copy_kernel dtb_file dtb

and no_dtb_dir kernel_name =
  eprintf "\
supermin: failed to find a dtb (device tree) directory.

I expected to take '%s' and to
replace vmlinuz- with dtb- to form a directory.

You can set SUPERMIN_KERNEL, SUPERMIN_MODULES and SUPERMIN_DTB
to override automatic selection.  See supermin(1).\n"
    kernel_name;
  exit 1

and no_dtb dtb_dir wildcard =
  eprintf "\
supermin: failed to find a matching device tree.

I looked for a file matching '%s' in directory '%s'.

You can set SUPERMIN_KERNEL, SUPERMIN_MODULES and SUPERMIN_DTB
to override automatic selection.  See supermin(1).\n"
    wildcard dtb_dir;
  exit 1

and find_modpath debug kernel_version =
  try
    let modpath = getenv "SUPERMIN_MODULES" in
    if debug >= 1 then
      printf "supermin: kernel: SUPERMIN_MODULES environment variable = %s\n%!"
        modpath;
    modpath
  with Not_found ->
    let modpath = "/lib/modules/" ^ kernel_version in
    if debug >= 1 then
      printf "supermin: kernel: picked modules path %s\n%!" modpath;
    modpath

and has_modpath kernel_name =
  try
    let kv = get_kernel_version kernel_name in
    modules_dep_exists kv
  with
  | Not_found -> false

and get_kernel_version kernel_name =
  if string_prefix "vmlinuz-" kernel_name then (
    let kv = String.sub kernel_name 8 (String.length kernel_name - 8) in
    if modules_dep_exists kv then kv
    else get_kernel_version_from_name kernel_name
  ) else get_kernel_version_from_name kernel_name

and modules_dep_exists kv =
  try (lstat ("/lib/modules/" ^ kv ^ "/modules.dep")).st_kind = S_REG
  with Unix_error _ -> false

and get_kernel_version_from_name kernel_name =
  get_kernel_version_from_file ("/boot" // kernel_name)

(* Extract the kernel version from a Linux kernel file.
 *
 * Returns a string containing the version or [Not_found] if the
 * file can't be read, is not a Linux kernel, or the version can't
 * be found.
 *
 * See ftp://ftp.astron.com/pub/file/file-<ver>.tar.gz
 * (file-<ver>/magic/Magdir/linux) for the rules used to find the
 * version number:
 *   514             string  HdrS     Linux kernel
 *   >518            leshort >0x1ff
 *   >>(526.s+0x200) string  >\0      version %s,
 *
 * Bugs: probably limited to x86 kernels.
 *)
and get_kernel_version_from_file file =
  try
    let chan = open_in file in
    let buf = read_string chan 514 4 in
    if buf <> "HdrS" then (
      close_in chan;
      raise Not_found
    );
    let s = read_leshort chan 518 in
    if s < 0x1ff then (
      close_in chan;
      raise Not_found
    );
    let offset = read_leshort chan 526 in
    if offset < 0 then (
      close_in chan;
      raise Not_found
    );
    let buf = read_string chan (offset + 0x200) 132 in
    close_in chan;
    let rec loop i =
      if i < 132 then (
        if buf.[i] = '\000' || buf.[i] = ' ' ||
          buf.[i] = '\t' || buf.[i] = '\n' then
          String.sub buf 0 i
        else
          loop (i+1)
      )
      else raise Not_found
    in
    loop 0
  with
  | Sys_error _ -> raise Not_found
  | Invalid_argument _ -> raise Not_found

(* Read an unsigned little endian short at a specified offset in a file. *)
and read_leshort chan offset =
  let buf = read_string chan offset 2 in
  (Char.code buf.[1] lsl 8) lor Char.code buf.[0]

and read_string chan offset len =
  seek_in chan offset;
  let buf = String.create len in
  really_input chan buf 0 len;
  buf

and copy_or_symlink_file copy_kernel src dest =
  if not copy_kernel then
    symlink src dest
  else (
    let cmd = sprintf "cp -p %s %s" (quote src) (quote dest) in
    run_command cmd
  )
